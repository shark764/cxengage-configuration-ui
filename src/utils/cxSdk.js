export function cxInit() {
  return CxEngage.initialize(require('../sdkConfig.json'));
}

export function cxLogin(username, password, callback) {
  CxEngage.authentication.login({ 
    username: username, 
    password: password,
    // ttl is good to comment out to be able to test what 
    // happens when authentication expires...
    // ttl in seconds 
    // ttl: 60 
  }, (err, topic, response) => {
    if (err) {
      console.warn('ERROR WITH LOGIN', err);
      return callback(err);
    } else {
      return callback(response);
    }
  });
}

export function cxTokenLogin(token, callback) {
  CxEngage.authentication.login({ token: token }, (err, topic, response) => {
    if (err) {
      console.warn('ERROR WITH LOGIN', err);
      return callback(err);
    } else {
      return callback(response);
    }
  });
}

export function cxSetTenant(selectedTenant, callback) {
  CxEngage.session.setActiveTenant({ tenantId: selectedTenant, noSession: true }, (err, topic, response) => {
    if (err) {
      console.warn('ERROR WITH TENANT SELECTION', err);
    } else {
      return callback(response);
    }
  });
}
