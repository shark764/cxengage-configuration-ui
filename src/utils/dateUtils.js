export const dateToString = date =>
  new Date(date).toLocaleString('en-US', {
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: true
  });
