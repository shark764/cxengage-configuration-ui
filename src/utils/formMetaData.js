export const createFormMetadata = {
  lists: ['listTypes'],
  outboundIdentifiers: ['flows']
};

export const updateFormMetadata = {
  outboundIdentifiers: ['flows']
};
