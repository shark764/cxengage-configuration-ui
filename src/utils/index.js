export const detailHeaderText = (itemArray, name) => {
  if (itemArray) {
    return `${itemArray.length > 1 ? itemArray.length : ''} ${name}`;
  } else {
    return name;
  }
};

export const parentUrl = document.referrer;

const cloneObject = obj => Object.assign({}, obj);

export const renameObjectKey = (object, key, newKey) => {
  const clonedObj = cloneObject(object);
  const targetKey = clonedObj[key];
  delete clonedObj[key];
  clonedObj[newKey] = targetKey;
  return clonedObj;
};

// Sort method to sort strings in entity tables
export const customSortStringMethod = (a, b) => {
  let tempArray = [];
  tempArray = [a, b];

  const collator = new Intl.Collator(undefined, { numeric: true, sensitivity: 'base' });
  tempArray.sort(collator.compare);

  if (tempArray[0] === a) {
    return -1;
  } else {
    return 1;
  }
};