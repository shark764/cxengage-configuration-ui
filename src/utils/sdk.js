import { isInIframe } from 'serenova-js-utils/browser';
import { capitalizeFirstLetter } from 'serenova-js-utils/strings';
import store from '../redux/store';
import { getCurrentEntity } from '../redux/modules/entities/selectors';
import { updateRealtimeStatisticsBatchData } from '../redux/modules/reporting/agentStateMonitoring';
import { updateTableData } from '../redux/modules/reporting/interactionMonitoring';

function guid() {
  function s4() {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
}

export const sdkPromise = sdkCall => {
  if (isInIframe()) {
    return new Promise((resolve, reject) => {
      const currentEntity = getCurrentEntity(store.getState());
      if (
        sdkCall.module === 'subscribe' &&
        (currentEntity === 'interactionMonitoring' || currentEntity === 'agentStateMonitoring')
      ) {
        CxEngage.subscribe(sdkCall.command, function(error, topic, response) {
          if (error) {
            console.error('ERROR', error);
            reject(error);
          } else {
            if (currentEntity === 'agentStateMonitoring') {
              store.dispatch(
                updateRealtimeStatisticsBatchData({
                  resources: response.resourceCapacity.body.results.resourceCapacity,
                  agentStates: response.resourceStateList.body.results.json
                })
              );
            } else {
              store.dispatch(updateTableData(response.interactionsInConversationList.body.results.interactions));
            }
            resolve(response);
          }
        });
      } else {
        /* istanbul ignore next */
        function handleCall(error, topic, response) {
          console.log('[SDK] SDK sending back:', error, topic, response);
          if (error) {
            console.warn('ERROR', error);
            if (error.data && error.data.apiResponse && error.data.apiResponse.status === 401) {
              store.dispatch({ type: 'TOGGLE_USER_AUTH' });
            }
            reject(error);
          } else {
            resolve(response);
          }
        }

        if (CxEngage[sdkCall.module][sdkCall.command] === undefined) {
          const requestObject = {
            path: sdkCall.path,
            body: sdkCall.data,
            customTopic: sdkCall.topic
          };
          if (sdkCall.apiVersion) {
            requestObject['apiVersion'] = sdkCall.apiVersion;
          }
          CxEngage.api[sdkCall.crudAction](requestObject, handleCall);
        } else {
          CxEngage[sdkCall.module][sdkCall.command](sdkCall.data, handleCall);
        }
      }
    });
  } else {
    return new Promise((resolve, reject) => {
      const completeSdkCall = { messageId: guid(), ...sdkCall };
      const handleResponse = event => {
        if (completeSdkCall.messageId === event.data.messageId) {
          const { error, response } = event.data;
          if (error) {
            reject(error);
          } else {
            resolve(response);
          }
          removeEventListener('message', handleResponse, false);
        }
      };
      addEventListener('message', handleResponse);
      window.parent.postMessage(completeSdkCall, '*');
    });
  }
};

export const sdkCall = sdkCall =>
  new Promise((resolve, reject) => {
    window.parent.postMessage(sdkCall, '*');
    resolve('Posted Message To Parent iFrame');
  });

export const errorLabel = error => {
  let errorDetails;
  if (
    error.data &&
    error.data.apiResponse &&
    error.data.apiResponse.apiResponse &&
    error.data.apiResponse.apiResponse.response &&
    error.data.apiResponse.apiResponse.response.error
  ) {
    let responseError = error.data.apiResponse.apiResponse.response.error;
    let { code, message } = responseError;
    errorDetails = ` ${code}: ${message}`;
    if (code === undefined || message === undefined) {
      errorDetails = ` ${error.data.apiResponse.status}: ${responseError}`;
    }
  }
  return `${error.message} ${errorDetails ? errorDetails : ''}`;
};

//errorManager
export const errorManager = error => {
  let messageFromAPI;
  let errorDetails;
  let attr;
  if (
    error.data &&
    error.data.apiResponse &&
    ((error.data.apiResponse.apiResponse &&
      error.data.apiResponse.apiResponse.response &&
      error.data.apiResponse.apiResponse.response.error) ||
      (error.data.apiResponse.response && error.data.apiResponse.response.error))
  ) {
    let code = error.data.apiResponse.status;
    let responseError =
      (error.data.apiResponse &&
        error.data.apiResponse.apiResponse &&
        error.data.apiResponse.apiResponse.response.error) ||
      error.data.apiResponse.response.error;
    if (responseError instanceof Array) {
      messageFromAPI = responseError[0];
    } else if (typeof responseError === 'object') {
      // Not all APIs return the same error structure
      // in some cases message is null or undefined, so
      // we add the same default message as Config-UI1.
      let { code, message = 'Some bulk actions could not be applied.', attribute } = responseError;
      attr = attribute && Object.keys(attribute)[0];
      errorDetails = ` ${code}: ${message}`;

      if (attribute && typeof attribute === 'object') {
        messageFromAPI = attribute[attr];
        errorDetails = ` ${error.message} ${code}: ${capitalizeFirstLetter(attribute[attr])}`;
      } else if (error.message && message) {
        messageFromAPI = message;
        errorDetails = ` ${error.message}<br/>${code}: ${capitalizeFirstLetter(message)}.`;
      } else if (code === undefined || message === undefined) {
        errorDetails = ` ${error.data.apiResponse.status}: ${responseError}`;
        messageFromAPI = message;
      }
    }

    if (
      ['same value', 'already exist', 'already registered', 'duplicate found'].some(function(keyword) {
        // checks whether a keyword exist in the API message
        return messageFromAPI.includes(keyword);
      }) &&
      // The only entity that has its own messages is DispatchMappings
      !messageFromAPI.includes('channel type')
    ) {
      errorDetails = ` ${
        error.message
      } ${code}: Resource with the same name or value already exists in the system, please enter a different value.`;
    }
  }
  return { errorMessage: errorDetails ? errorDetails : 'An error has occurred.', attribute: attr };
};
