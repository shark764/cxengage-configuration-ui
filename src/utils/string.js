export const capitalizeFirstLetter = string =>
  `${string.charAt(0).toUpperCase()}${string.slice(1)}`;

export const removeLastLetter = string =>
  string.substring(0, string.length - 1);

export const camelCaseToKebabCase = string =>
  string.replace(/([a-z][A-Z])/g, function(match) {
    return match.substr(0, 1) + '-' + match.substr(1, 1).toLowerCase();
  });
