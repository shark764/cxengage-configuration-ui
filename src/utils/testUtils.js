/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import { createStore } from 'redux';

export const mockStore = createStore(state => state);
