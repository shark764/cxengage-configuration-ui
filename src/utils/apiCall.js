import { sdkPromise } from './sdk';

export function apiCall(urlSuffix, method = 'GET', data) {
  return sdkPromise({
    module: 'updateLocalStorage',
    command: `updateLocalStorage`,
    data: 'updateLocalStorage',
    topic: 'updateLocalStorage'
  }).then(({tenant: {tenantId}, token, baseUrl}) => {
    const url = `${baseUrl}/v1/tenants/${tenantId}/${urlSuffix}`;
    const requestParams =  {
      mode: "cors", // no-cors, cors, *same-origin
      cache: "no-cache", // *default, no-cache, reload, force-cache, only-if-cached
      Authorization: 'Token ' + token,
      headers: {
          "Content-Type": "application/json",
      },
      redirect: "follow", // manual, *follow, error
      referrer: "no-referrer", // no-referrer, *client
    };
    requestParams.method = method;
    if(token) {
      requestParams.headers.Authorization = 'Token ' + token;
    }
    if(data) {
      requestParams.body = JSON.stringify(data);
    }
      return fetch(url, requestParams)
      .then(response => {
        
        if(response.status === 401) {
          return 'Unauthorized';
        } else if ( response.status >= 300) {
          return;
        } else {
          return response.json();
        }
      });

  });
}

export function updateBetaFeatures(featuresObject) {
  return apiCall('settings/betaFeatures/value', 'PUT', featuresObject);
}

export function readBetaFeatures(featuresObject) {
  return apiCall('settings/betaFeatures/value');
}