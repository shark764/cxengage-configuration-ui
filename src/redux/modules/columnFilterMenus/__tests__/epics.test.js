// /*
//  * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
//  */

import { ActionsObservable } from 'redux-observable';
import { mockStore } from 'TestUtils';
import { sdkPromise } from '../../../../utils/sdk';

import {
  SaveColumnsToLocalStorage,
  UpdateStatSubscriptionFilters,
  UpdateSkillsAndGroupsFilter
} from '../epics';

import {
  selectInteractionMonitoringColumns,
  selectGroups,
  selectSkills
} from '../selectors';

jest.mock('../../../../utils/sdk');
jest.mock('../selectors');
selectInteractionMonitoringColumns.mockReturnValue('mockColumnData');
selectSkills.mockReturnValue([
  { active: true, skillId: '001' },
  { active: true, skillId: '002' },
  { active: false, skillId: '003' }
]);
selectGroups.mockReturnValue([
  { active: true, groupId: '001' },
  { active: true, groupId: '002' },
  { active: false, groupId: '003' }
]);

describe('SaveColumnsToLocalStorage', () => {
  it('localstorage is updated on column changes', done => {
    const action = ActionsObservable.of({
      type: 'TOGGLE_MENUITEMS',
      menuType: 'Columns'
    });
    SaveColumnsToLocalStorage(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
});

describe('UpdateSkillsAndGroupsFilter', () => {
  it('localstorage is updated on skills column changes', done => {
    const action = ActionsObservable.of({
      type: 'FETCH_DATA_FULFILLED',
      entityName: 'skills',
      response: { result: 'results' },
      tableType: 'mockTableSkills'
    });
    UpdateSkillsAndGroupsFilter(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('localstorage is updated on groups column changes', done => {
    const action = ActionsObservable.of({
      type: 'FETCH_DATA_FULFILLED',
      entityName: 'groups',
      response: { result: 'results' },
      tableType: 'mockTableGroups'
    });
    UpdateSkillsAndGroupsFilter(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
});

describe('UpdateStatSubscriptionFilters', () => {
  afterEach(() => {
    sdkPromise.mockClear();
  });
  it('dispatch the proper action on success: skills / empty array', done => {
    selectSkills.mockReturnValueOnce([]);
    selectGroups.mockReturnValueOnce([]);
    sdkPromise.mockReturnValue(
      new Promise(resolve => resolve('mock response'))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Skills'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('dispatch the proper action on success: groups / empty array', done => {
    selectSkills.mockReturnValueOnce([]);
    selectGroups.mockReturnValueOnce([]);
    sdkPromise.mockReturnValue(
      new Promise(resolve => resolve('mock response'))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Groups'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('dispatch the proper action on success: skills', done => {
    sdkPromise.mockReturnValue(
      new Promise(resolve => resolve('mock response'))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Skills'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('dispatch the proper action on success: groups', done => {
    sdkPromise.mockReturnValue(
      new Promise(resolve => resolve('mock response'))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Groups'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('dispatch the proper action on failure: skills', done => {
    sdkPromise.mockReturnValue(
      new Promise((resolve, reject) => reject(new Error()))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Skills'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
  it('dispatch the proper action on failure: groups', done => {
    sdkPromise.mockReturnValue(
      new Promise((resolve, reject) => reject(new Error()))
    );
    const action = ActionsObservable.of({
      type: 'TOGGLE_ALL_MENUITEMS_OFF',
      menuType: 'Groups'
    });
    UpdateStatSubscriptionFilters(action, mockStore).subscribe(
      actualOutputActions => {
        expect(actualOutputActions).toMatchSnapshot();
        done();
      }
    );
  });
});
