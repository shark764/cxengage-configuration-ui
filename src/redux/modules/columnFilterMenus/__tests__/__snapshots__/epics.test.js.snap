// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveColumnsToLocalStorage localstorage is updated on column changes 1`] = `
Object {
  "columnsData": "\\"mockColumnData\\"",
  "type": "COLUMNS_LOCALSTORAGE_UPDATED_$",
}
`;

exports[`UpdateSkillsAndGroupsFilter localstorage is updated on groups column changes 1`] = `
Object {
  "arrayOfGroupData": "results",
  "tableType": "mockTableGroups",
  "type": "SET_GROUPS_DATA",
}
`;

exports[`UpdateSkillsAndGroupsFilter localstorage is updated on skills column changes 1`] = `
Object {
  "arrayOfSkillData": "results",
  "tableType": "mockTableSkills",
  "type": "SET_SKILLS_DATA",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on failure: groups 1`] = `
Object {
  "type": "STATS_NOT_UPDATED_$",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on failure: skills 1`] = `
Object {
  "type": "STATS_NOT_UPDATED_$",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on success: groups / empty array 1`] = `
Object {
  "type": "STATS_UPDATED_$",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on success: groups 1`] = `
Object {
  "type": "STATS_UPDATED_$",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on success: skills / empty array 1`] = `
Object {
  "type": "STATS_UPDATED_$",
}
`;

exports[`UpdateStatSubscriptionFilters dispatch the proper action on success: skills 1`] = `
Object {
  "type": "STATS_UPDATED_$",
}
`;
